Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -0,0 +1,235 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+
+! This file was ported from Lean 3 source module algebra.group.unique_prods
+! leanprover-community/mathlib commit d6fad0e5bf2d6f48da9175d25c3dc5706b3834ce
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Finset.Preimage
+
+/-!
+#  Unique products and related notions
+
+A group `G` has *unique products* if for any two non-empty finite subsets `A, B ⊂ G`, there is an
+element `g ∈ A * B` that can be written uniquely as a product of an element of `A` and an element
+of `B`.  We call the formalization this property `UniqueProds`.  Since the condition requires no
+property of the group operation, we define it for a Type simply satisfying `Mul`.  We also
+introduce the analogous "additive" companion, `UniqueSums` and link the two so that `to_additive`
+converts `UniqueProds` into `UniqueSums`.
+
+Here you can see several examples of Types that have `UniqueSums/Prods`
+(`apply_instance` uses `covariants.to_uniqueProds` and `covariants.to_uniqueSums`).
+```lean
+import data.real.basic
+
+example : unique_sums ℕ   := by apply_instance
+example : unique_sums ℕ+  := by apply_instance
+example : unique_sums ℤ   := by apply_instance
+example : unique_sums ℚ   := by apply_instance
+example : unique_sums ℝ   := by apply_instance
+example : unique_prods ℕ+ := by apply_instance
```

Comment:
Please Lean4-ify

Reply:
I have fixed the example and made sure it works :)